src/coreclr/jit/lsraarm64.cpp

```diff
@@ -25,6 +25,415 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+            {
+                continue;
+            }
+
+            return false;
+        }
+    } while (++i != registersCount);
+
+    return true;
+}
+
+//------------------------------------------------------------------------
+// filterConsecutiveCandidates: Given `candidates`, check if `registersNeeded` consecutive
+//   registers are available in it, and if yes, returns first bit set of every possible series.
+//
+// Arguments:
+//    candidates                - Set of availble candidates.
+//    registersNeeded           - Number of consecutive registers needed.
+//    allConsecutiveCandidates  - Mask returned containing all bits set for possible consecutive register candidates.
+//
+//  Returns:
+//      From `candidates`, the mask of series of consecutive registers of `registersNeeded` size with just the first-bit
+//      set.
+//
+regMaskTP LinearScan::filterConsecutiveCandidates(regMaskTP    candidates,
+                                                  unsigned int registersNeeded,
+                                                  regMaskTP*   allConsecutiveCandidates)
+{
+    if (BitOperations::PopCount(candidates) < registersNeeded)
+    {
+        // There is no way the register demanded can be satisfied for this refposition
+        // based on the candidates from which it can allocate a register.
+        return RBM_NONE;
+    }
+
+    regMaskTP currAvailableRegs      = candidates;
+    regMaskTP overallResult          = RBM_NONE;
+    regMaskTP consecutiveResult      = RBM_NONE;
+    regMaskTP busyRegsInThisLocation = regsBusyUntilKill | regsInUseThisLocation;
+
+// At this point, for 'n' registers requirement, if Rm, Rm+1, Rm+2, ..., Rm+k-1 are
+// available, create the mask only for Rm, Rm+1, ..., Rm+(k-n) to convey that it
+// is safe to assign any of those registers, but not beyond that.
+#define AppendConsecutiveMask(startIndex, endIndex, availableRegistersMask)                                            \
+    regMaskTP selectionStartMask = (1ULL << regAvailableStartIndex) - 1;                                               \
+    regMaskTP selectionEndMask   = (1ULL << (regAvailableEndIndex - registersNeeded + 1)) - 1;                         \
+    consecutiveResult |= availableRegistersMask & (selectionEndMask & ~selectionStartMask);                            \
+    overallResult |= availableRegistersMask;
+
+    DWORD regAvailableStartIndex = 0, regAvailableEndIndex = 0;
+
+    // If we don't find consecutive registers, also track which registers we can pick so
+    // as to reduce the number of registers we will have to spill, to accomodate the
+    // request of the consecutive registers.
+    regMaskTP registersNeededMask = (1ULL << registersNeeded) - 1;
+
+    do
+    {
+        // From LSB, find the first available register (bit `1`)
+        BitScanForward64(&regAvailableStartIndex, static_cast<DWORD64>(currAvailableRegs));
+        regMaskTP startMask = (1ULL << regAvailableStartIndex) - 1;
```

Comment:
Can `regAvailableStartIndex` ever be `REG_V31` (aka, 63)? Is this expression well-defined in that case?

Reply:
Yes and it gets handled by `if (maskProcessed == 0)` case.